programs/drift/src/controller/pnl.rs

```diff
@@ -129,20 +131,46 @@ pub fn settle_pnl(
     let perp_market = &mut perp_market_map.get_ref_mut(&market_index)?;
 
     if perp_market.amm.curve_update_intensity > 0 {
-        validate!(
-            perp_market.amm.last_oracle_valid,
-            ErrorCode::InvalidOracle,
-            "Oracle Price detected as invalid"
-        )?;
+        let healthy_oracle = perp_market.amm.last_oracle_valid
+            && oracle_map.slot == perp_market.amm.last_update_slot;
+        if !healthy_oracle {
+            let (_oracle_price_data, oracle_validity) = oracle_map.get_price_data_and_validity(
+                &perp_market.amm.oracle,
+                perp_market
+                    .amm
+                    .historical_oracle_data
+                    .last_oracle_price_twap,
+            )?;
 
-        validate!(
-            oracle_map.slot == perp_market.amm.last_update_slot,
-            ErrorCode::AMMNotUpdatedInSameSlot,
-            "AMM must be updated in a prior instruction within same slot (current={} != amm={}, last_oracle_valid={})",
-            oracle_map.slot,
-            perp_market.amm.last_update_slot,
-            perp_market.amm.last_oracle_valid
-        )?;
+            let oracle_divergence = oracle_price
+                .safe_sub(
+                    perp_market
+                        .amm
+                        .historical_oracle_data
+                        .last_oracle_price_twap_5min,
+                )?
+                .safe_mul(PERCENTAGE_PRECISION_I64)?
+                .safe_div(oracle_price)?;
+
+            if is_oracle_valid_for_action(oracle_validity, Some(DriftAction::SettlePnl))?
+                && oracle_divergence < PERCENTAGE_PRECISION_I64 / 200
```

Comment:
should this be based on some per market variable like spread or s.t.?

Reply:
added by contract tier